lib/events/athena/consumer.go

```diff
@@ -0,0 +1,439 @@
+	visibilityTimeout     int32
+	waitOnReceiveTimeout  int32
+	waitOnReceiveDuration time.Duration
+
+	batchMaxItems int
+}
+
+type sqsCli interface {
+	ReceiveMessage(ctx context.Context, params *sqs.ReceiveMessageInput, optFns ...func(*sqs.Options)) (*sqs.ReceiveMessageOutput, error)
+}
+
+type s3downloader interface {
+	Download(ctx context.Context, w io.WriterAt, input *s3.GetObjectInput, options ...func(*manager.Downloader)) (n int64, err error)
+}
+
+func newConsumer(cfg Config, awsCfg aws.Config, backend backend.Backend, logEntry *log.Entry) (*consumer, error) {
+	u, err := url.Parse(cfg.LocationS3)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	largeEventsURL, err := url.Parse(cfg.LargeEventsS3)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s3cli := s3.NewFromConfig(awsCfg)
+	sqsCli := sqs.NewFromConfig(awsCfg)
+	return &consumer{
+		Entry:               logEntry,
+		sqsCli:              sqsCli,
+		backend:             backend,
+		storeLocationPrefix: strings.TrimSuffix(strings.TrimPrefix(u.Path, "/"), "/"),
+		storeLocationBucket: u.Host,
+		batchMaxItems:       cfg.BatchMaxItems,
+		batchMaxInterval:    cfg.BatchMaxInterval,
+		collectConfig: sqsCollectConfig{
+			sqsCli:   sqsCli,
+			queueURL: cfg.QueueURL,
+			// TODO(tobiaszheller): use s3 manager from teleport observability.
+			payloadDownloader:     manager.NewDownloader(s3cli),
+			payloadBucket:         largeEventsURL.Host,
+			visibilityTimeout:     int32(cfg.BatchMaxInterval.Seconds()),
+			waitOnReceiveTimeout:  int32(maxWaitTimeOnReceiveMessageFromSQS.Seconds()),
+			waitOnReceiveDuration: maxWaitTimeOnReceiveMessageFromSQS,
+			batchMaxItems:         cfg.BatchMaxItems,
+		},
+	}, nil
+}
+
+// run continuously runs batching job. It is blocking operation.
+// It is stopped via canceling context.
+func (c *consumer) run(ctx context.Context) {
+	const lockName = "athena_batcher"
+	// lockTTL is time after which other auth server can acquire lock.
+	// It can be half of BatchMaxInterval because RunWhileLocked automatically
+	// refresh TTL if job is still running.
+	lockTTL := c.batchMaxInterval / 2
+	var err error
+	for {
+		err = backend.RunWhileLocked(ctx, c.backend, lockName, lockTTL, func(ctx context.Context) error {
+			return trace.Wrap(c.singleBatch(ctx))
+		})
+		if err != nil {
+			// Ctx.Cancel is used to stop batcher
+			if ctx.Err() != nil {
+				c.Debug("Batcher has been stopped")
+				return
+			}
+			c.Errorf("Batcher single run failed: %v", err)
+		}
+	}
+}
+
+// singleBatch creates single batch of events. It waits either up to BatchMaxInterval
+// or BatchMaxItems while reading events from queue. Batch is sent to s3 as
+// parquet file and at the end events are deleted from queue.
+func (c *consumer) singleBatch(ctx context.Context) error {
+	start := time.Now()
+	var size int
+	// TODO(tobiaszheller): we need some metrics to track it.
+	defer func() {
+		c.Debugf("Batch of %d took: %s", size, time.Since(start))
```

Comment:
It will print it every minute, nevertheless I left that todo on purpose. I am planning to come back to it in next PR when adding metrics to it and remove that log line.

Reply:
done